Stroke prediction dataset (@Fedesoriano on Kaggle)
Here, we will try to implement a supervised learning algorithm to predict the probability of someone to have a stroke or not
We will use Decision Tree for classification

In [14]:
#Load libraries
import numpy as np
import pandas as pd
from sklearn.metrics import f1_score,accuracy_score
from sklearn.tree import DecisionTreeClassifier
from sklearn.model_selection import train_test_split

In [15]:
#Loading the Dataset
df_stroke = pd.read_csv('Stroke_Prediction_Dataset_from_Fedesoriano.csv')
df_stroke

Unnamed: 0     id  gender   age  hypertension  heart_disease  \
0              0   9046    Male  67.0             0              1   
1              1  51676  Female  61.0             0              0   
2              2  31112    Male  80.0             0              1   
3              3  60182  Female  49.0             0              0   
4              4   1665  Female  79.0             1              0   
...          ...    ...     ...   ...           ...            ...   
5105        5105  18234  Female  80.0             1              0   
5106        5106  44873  Female  81.0             0              0   
5107        5107  19723  Female  35.0             0              0   
5108        5108  37544    Male  51.0             0              0   
5109        5109  44679  Female  44.0             0              0   

     ever_married      work_type Residence_type  avg_glucose_level   bmi  \
0             Yes        Private          Urban             228.69  36.6   
1             Yes  Self-employed          Rural             202.21   NaN   
2             Yes        Private          Rural             105.92  32.5   
3             Yes        Private          Urban             171.23  34.4   
4             Yes  Self-employed          Rural             174.12  24.0   
...           ...            ...            ...                ...   ...   
5105          Yes        Private          Urban              83.75   NaN   
5106          Yes  Self-employed          Urban             125.20  40.0   
5107          Yes  Self-employed          Rural              82.99  30.6   
5108          Yes        Private          Rural             166.29  25.6   
5109          Yes       Govt_job          Urban              85.28  26.2   

       smoking_status  stroke  
0     formerly smoked       1  
1        never smoked       1  
2        never smoked       1  
3              smokes       1  
4        never smoked       1  
...               ...     ...  
5105     never smoked       0  
5106     never smoked       0  
5107     never smoked       0  
5108  formerly smoked       0  
5109          Unknown       0  

[5110 rows x 13 columns]

In [16]:
#Preprocessing the Data by selecting only 4 features (arbitrarily)
df_stroke = df_stroke[['age', 'hypertension', 'heart_disease', 'avg_glucose_level', 'bmi', 'stroke']]
df_stroke = df_stroke.dropna(subset=['bmi'], axis=0)
df_stroke

age  hypertension  heart_disease  avg_glucose_level   bmi  stroke
0     67.0             0              1             228.69  36.6       1
2     80.0             0              1             105.92  32.5       1
3     49.0             0              0             171.23  34.4       1
4     79.0             1              0             174.12  24.0       1
5     81.0             0              0             186.21  29.0       1
...    ...           ...            ...                ...   ...     ...
5104  13.0             0              0             103.08  18.6       0
5106  81.0             0              0             125.20  40.0       0
5107  35.0             0              0              82.99  30.6       0
5108  51.0             0              0             166.29  25.6       0
5109  44.0             0              0              85.28  26.2       0

[4909 rows x 6 columns]

In [17]:
#Splitting into Train and Test set
y = df_stroke.stroke
df_stroke.drop(['stroke'], axis=1, inplace=True)
X = df_stroke

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.25, random_state=2)

In [18]:
#Training the model
DT = DecisionTreeClassifier()
DT.fit(X_train,y_train)

DecisionTreeClassifier()

In [19]:
#Evaluating the model
pred=DT.predict(X_test)

print(accuracy_score(y_test,pred))
print(f1_score(y_test,pred))

0.9161237785016286
0.13445378151260504


With DecisionTree algorithm, we can predict the probability of someone to have a stroke or not with an accuracy of 92 %